store/model.go

```diff
@@ -0,0 +1,276 @@
+		exists, err = txn.Has(id)
+		return err
+	})
+	return
+}
+
+type Txn struct {
+	model     *Model
+	discarded bool
+	commited  bool
+	readonly  bool
+
+	actions []es.Action
+}
+
+type SaveOp struct {
+	Before interface{}
+	After  interface{}
+}
+
+func (t *Txn) Create(new interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	id := getEntityID(new)
+	if id == es.EmptyEntityID {
+		id = setNewEntityID(new)
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return errCantCreateExistingInstance
+	}
+
+	a := es.Action{
+		Type:       es.Create,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    new,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Save(updated interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+
+	id := getEntityID(updated)
+	key := t.model.dsKey.ChildString(id.String())
+	beforeBytes, err := t.model.datastore.Get(key)
+	if err == ds.ErrNotFound {
+		return errCantSaveNonExistentInstance
+	}
+	if err != nil {
+		return err
+	}
+
+	before := reflect.New(t.model.valueType.Elem()).Interface()
+	err = json.Unmarshal(beforeBytes, before)
+	if err != nil {
+		return err
+	}
+	a := es.Action{
+		Type:       es.Save,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   before,
+		Current:    updated,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Delete(id es.EntityID) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return ErrNotFound
+	}
+	a := es.Action{
+		Type:       es.Delete,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    nil,
+	}
+	t.actions = append(t.actions, a)
+	return nil
+}
+
+func (t *Txn) Has(id es.EntityID) (bool, error) {
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return false, err
+	}
+	return exists, nil
+}
+
+func (t *Txn) FindByID(id es.EntityID, v interface{}) error {
+	key := t.model.dsKey.ChildString(id.String())
+	bytes, err := t.model.datastore.Get(key)
+	if errors.Is(err, ds.ErrNotFound) {
+		return ErrNotFound
+	}
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(bytes, v)
+}
+
+func (t *Txn) Commit() error {
+	if t.discarded || t.commited {
+		return errAlreadyDiscardedCommitedTxn
+	}
+
+	events, err := t.model.eventcreator.Create(t.actions)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range events {
+		if err := t.model.dispatcher.Dispatch(e); err != nil {
+			return err // ToDo/Note: Important to document the implications of a partial dispatch
+		}
+	}
+	return nil
+}
+
+func (m *Model) Reduce(event es.Event) error {
+	// ToDo: distinguish local and remote events for proper locking
```

Comment:
Super important comment here that I'd like some opinions.

On every Txn start, a mutex is locked for the Model, in order to serialize changes and let the txn logic be sure that not other things are happening until after commits. Simple.

But we have a tricky situation. When the `Commit` is done, the generated `Events` are dispatcher to the `Dispatcher`, which indirectly calls the model `Reducer` which really applies the changes to the datastore.

The problem is that a call to `Reduce` here, can't be distinguished from a local committed transaction, or a remote event that came from the network... so adding in the quoted line a `lock.Lock()` can't be done. If the call to the `Reduce` is from a txn `Commit`, then the lock is already locked. 

The first thing that cames to my mind is to change the `Reduce` parameters to add an extra `local bool` parameter. If the `Dispatcher` sent `true` it means that we don't need to `Lock()` the mutex since implicitly means that is already locked. If it is `false`, it means that the `Event` sent from the `Dispatcher` isn't from a local txn but received from the network, so it's safe to `Lock()` to not interfere with local action.

Hope this explanation makes sense... if not, just ask :smile:

Reply:
Flash insight: The solution is actually more simple (since above one is plain incorrect since can have deadlocks).

Write transactions should acquire a write lock of the `Dispatcher`. Read transactions should acquire a read lock of the `Dispatcher`. Saying it differently, a txn is `Dispatcher`-wise, not model wise, since every model `Commit` can affect other models too (since `Dispatcher` dispatch events to all registered reducer and thus different models).

TL;DR: A `WriteTxn` is a stop-the-world thing to let the logic do things safely. Only one `WriteTxn` can exist at anytime in all models registered in the `Dispatcher`. The good thing about having `ReadTxn` is that multiple ones can be running at the same time. Any *remote* events dispatched remotely (because new events from logs where discovered) would be like a `WriteTxn`.

Just writing this because it popped out in my mind now.

Reply:
Lots of wires in here! 😅 I like the consideration of keeping order where possible.

Reply:
Agreed, if we can preserve arrival ordering, we're less reliant on CRDT structures down the line if we don't want them. I think your flash insight is on the money for sure. Seems reasonable. Let's go with that for now, and see how it plays out?

Reply:
Right, so I'll inclined towards a safe in-order approach, and safeguard all models state registered in the `Dispatcher`. If appears to be a problem in the future, we can have another round of dicussion!